oap-server/server-query-plugin/query-graphql-plugin/src/main/java/org/apache/skywalking/oap/query/graphql/resolver/AlarmQuery.java

```diff
@@ -62,11 +83,111 @@ public Alarms getAlarm(final Duration duration, final Scope scope, final String
+                        a.setEvents(sourceServiceEvent);
+                    }
+                    if (CollectionUtils.isNotEmpty(destServiceEvent)) {
+                        a.getEvents().addAll(destServiceEvent);
+                    }
+                    break;
+                case DefaultScopeDefine.SERVICE_INSTANCE :
+                    IDManager.ServiceInstanceID.InstanceIDDefinition instanceIDDefinition = IDManager.ServiceInstanceID.analysisId(a.getId());
+                    String serviceInstanceName = instanceIDDefinition.getName();
+                    String serviceName = IDManager.ServiceID.analysisId(instanceIDDefinition.getServiceId()).getName();
+                    List<Event> serviceInstanceEvent = mappingMap.get(serviceName);
+                    if (CollectionUtils.isNotEmpty(serviceInstanceEvent)) {
+                        List<Event> filterEvents = serviceInstanceEvent.stream().filter(e -> StringUtils.equals(e.getSource().getServiceInstance(), serviceInstanceName)).collect(Collectors.toList());
+                        a.setEvents(filterEvents);
+                    }
+                    break;
+                case DefaultScopeDefine.SERVICE_INSTANCE_RELATION :
+                    IDManager.ServiceInstanceID.InstanceIDDefinition sourceInstanceIDDefinition = IDManager.ServiceInstanceID.analysisId(a.getId());
+                    String sourceServiceInstanceName = sourceInstanceIDDefinition.getName();
+                    String sourceServiceName = IDManager.ServiceID.analysisId(sourceInstanceIDDefinition.getServiceId()).getName();
+
+                    IDManager.ServiceInstanceID.InstanceIDDefinition destInstanceIDDefinition = IDManager.ServiceInstanceID.analysisId(a.getId1());
+                    String destServiceInstanceName = destInstanceIDDefinition.getName();
+                    String destServiceName = IDManager.ServiceID.analysisId(destInstanceIDDefinition.getServiceId()).getName();
+
+                    List<Event> sourceInstanceEvent = mappingMap.get(sourceServiceName);
+                    List<Event> destInstanceEvent = mappingMap.get(destServiceName);
+
+                    if (CollectionUtils.isNotEmpty(sourceInstanceEvent)) {
+                        List<Event> filterEvents = sourceInstanceEvent.stream().filter(e -> StringUtils.equals(e.getSource().getServiceInstance(), sourceServiceInstanceName)).collect(Collectors.toList());
+                        a.setEvents(filterEvents);
+                    }
+                    if (CollectionUtils.isNotEmpty(destInstanceEvent)) {
+                        List<Event> filterEvents = destInstanceEvent.stream().filter(e -> StringUtils.equals(e.getSource().getServiceInstance(), destServiceInstanceName)).collect(Collectors.toList());
+                        a.getEvents().addAll(filterEvents);
+                    }
+                    break;
+                case DefaultScopeDefine.ENDPOINT :
```

Comment:
As for the `endpoint` scope, I'm wondering what do you think about the idea that we simply fallback to the service instance level events, because there are rarely endpoint events in my opinion, but endpoint alarms are highly possible triggered by instance events. Associating endpoint alarm with endpoint events ONLY may be correct, but it's highly possible to miss the relationship between them once there missed the endpoint events.

Reply:
Copy,  Thanks for your answer.

Reply:
@kezhenxu94 I think currently, I can't see a chance to see endpoint-level event, unless we have IDE or some code analysis tech to report there is logic changed in an endpoint scope.

> but endpoint alarms are highly possible triggered by instance events

I think you can't have endpoint-instance mapping, only could have service and service's all instances events, are you asking to get all of them?

Reply:
I need to determine the number and types of **case branches** again here. I am not sure if I understand it correctly, it seems that the `case` judgment of `ENDPOINT_RELATION` needs to be **removed** here.

Reply:
> I think you can't have endpoint-instance mapping, only could have service and service's all instances events.

OK I missed that.

> are you asking to get all of them?

If we can't get instance-endpoint mapping, listing all of the instances' events is one way, WDYT? @wu-sheng

Reply:
> we can't get instance-endpoint mapping

No, we can't. I prefer a service level event should be enough for the endpoint, after all, the endpoint belongs to a service. The unstable of a single instance can't impact the performance of an endpoint in most cases unless there is a load balance issue.